drivers/sx1276/sx1276.c

```diff
@@ -0,0 +1,1302 @@
+                    xtimer_remove(&dev->_internal.rx_timeout_timer);
+
+                    sx1276_set_status(dev,  SX1276_RF_IDLE);
+
+                    send_event(dev, SX1276_RX_TIMEOUT);
+                    break;
+                default:
+                    break;
+            }
+            break;
+        case SX1276_RF_TX_RUNNING:
+            break;
+        default:
+            break;
+    }
+}
+
+void sx1276_on_dio2(void *arg)
+{
+    /* Get interrupt context */
+    sx1276_t *dev = (sx1276_t *) arg;
+
+    switch (dev->settings.state) {
+        case SX1276_RF_RX_RUNNING:
+            switch (dev->settings.modem) {
+                case SX1276_MODEM_LORA:
+                    if (dev->settings.lora.freq_hop_on) {
+                        /* Clear IRQ */
+                        sx1276_reg_write(dev, SX1276_REG_LR_IRQFLAGS, SX1276_RF_LORA_IRQFLAGS_FHSSCHANGEDCHANNEL);
+
+                        dev->_internal.last_channel = sx1276_reg_read(dev, SX1276_REG_LR_HOPCHANNEL) & SX1276_RF_LORA_HOPCHANNEL_CHANNEL_MASK;
+                        send_event(dev, SX1276_FHSS_CHANGE_CHANNEL);
+                    }
+
+                    break;
+                default:
+                    break;
+            }
+            break;
+        case SX1276_RF_TX_RUNNING:
+            switch (dev->settings.modem) {
+                case SX1276_MODEM_FSK:
+                    break;
+                case SX1276_MODEM_LORA:
+                    if (dev->settings.lora.freq_hop_on) {
+                        /* Clear IRQ */
+                        sx1276_reg_write(dev, SX1276_REG_LR_IRQFLAGS, SX1276_RF_LORA_IRQFLAGS_FHSSCHANGEDCHANNEL);
+
+                        dev->_internal.last_channel = sx1276_reg_read(dev, SX1276_REG_LR_HOPCHANNEL) & SX1276_RF_LORA_HOPCHANNEL_CHANNEL_MASK;
+                        send_event(dev, SX1276_FHSS_CHANGE_CHANNEL);
+                    }
+                    break;
+                default:
+                    break;
+            }
+            break;
+        default:
+            break;
+    }
+}
+
+void sx1276_on_dio3(void *arg)
+{
+    /* Get interrupt context */
+    sx1276_t *dev = (sx1276_t *) arg;
+
+    switch (dev->settings.modem) {
+        case SX1276_MODEM_FSK:
+            break;
+        case SX1276_MODEM_LORA:
+            /* Clear IRQ */
+            sx1276_reg_write(dev, SX1276_REG_LR_IRQFLAGS, SX1276_RF_LORA_IRQFLAGS_CADDETECTED | SX1276_RF_LORA_IRQFLAGS_CADDONE);
+
+            /* Send event message */
+            dev->_internal.is_last_cad_success = (sx1276_reg_read(dev, SX1276_REG_LR_IRQFLAGS) & SX1276_RF_LORA_IRQFLAGS_CADDETECTED) == SX1276_RF_LORA_IRQFLAGS_CADDETECTED;
+            send_event(dev, SX1276_CAD_DONE);
+            break;
+        default:
+            break;
+    }
+}
+
+/* Following interrupt lines are not used */
+void sx1276_on_dio4(void *arg)
+{
+    (void) arg;
+}
+
+void sx1276_on_dio5(void *arg)
+{
+    (void) arg;
+}
+
+void *dio_polling_thread(void *arg)
```

Comment:
Hi! I'll try to clarify reasons why I did that.

The main point is to not doing complex stuff inside DIO's ISRs and do it in separate thread instead.

> Doesn't this make packet loss more likely?

I think no, the RX handling is pretty straightforward and doesn't evolve complex mechanisms such as packet queues which may overflow or something like that.

> Plus doesn't this implement a pooling system witch is cpu and energy intensive?

The dio_polling_thread _doesn't_ poll a DIO pins as busy-waiting one. The major part of a time this thread is blocked on awaiting _messages_ that comes from DIO interrupts.

As this thread is awaiting for messages in blocked state, it's not CPU-intensive at all.

> Also, is it really needed since you are using stack memory and other resources

This thread is needed since the handler of DIO interrupt might be pretty heavy in matter of stack usage. For example, consider the DIO0 interrupt on RX. After the RX event occurs, if we just call the event handler's callback and If we put some stack-heavy MAC logic into that callback, the ISR stack will definitely be overflowed than if we will do event callback from the separate thread.

The stack space is allocated for this thread only.

Another solution I came up with is to force user to create a threads to handle messages from the driver, but it's seems to me a clear violation of implementation hiding principle and that will lead to more ways to do something wrong.

I hope I made the rationale clear for you.

Cheers.